### ISSUE TITLE
[Memory Issue] tfLearn killing my net. 

### ISSUE THREAD

**Issue by gergf** [tflearn/tflearn#683]
First of all, thanks in advance for your time.

I have built a "small" conv net to do semantic segmentation. The problem is that, when I run the net, my laptop gets frozen, and tfLearns ends killing the process. I guess that I am doing something wrong, but I can't figure out what...

I've calculated the num of parameters and it's around 1500. I've also calculated the memory needed and I get around 200MB. I'm running the net in a laptop (4GB ram, 4GB swap, Intel(R) Core(TM) i3-4030U CPU @ 1.90GHz), so I should have no problems running this. 

TL;DR: My PC is using around 5GB (full RAM and 2GB in swap) when it's supposed to use only ~200MB... Any advice to debug my program?

This is the red that I'am running:
	
	print("Loading the data...")
	dataLoader = loader(corpusPath="/home/ger/Desktop/Seg/")
	batchSize=20
	# Load the images. This returns two nparrays (batchSize, 256,256)
	X, Y = dataLoader.getTrainBatch(batchSize=batchSize)
	X = np.reshape(X, (batchSize,256,256,1))
	Y = np.reshape(Y, (batchSize,256,256,1))
	# print(X.shape)

	# Build the nn #
	print("Building the model...")
	network = tflearn.input_data(shape=[None,256,256,1])

	# Encoder # 
	network = tflearn.conv_2d(network, nb_filter=64, filter_size=3, activation='relu', name='conv1', padding="SAME")
	network = tflearn.conv_2d(network, nb_filter=64, filter_size=3, activation='relu', name='conv2', padding="SAME")
	network = tflearn.max_pool_2d(network, 2)

	# Decoder # 
	network = tflearn.conv_2d_transpose(network, nb_filter=64, filter_size=3, strides=2, output_shape=[256,256], name='transConv1')
	network = tflearn.conv_2d(network, nb_filter=64, filter_size=3, activation='relu', name='conv3', padding="SAME")
	network = tflearn.conv_2d(network, nb_filter=64, filter_size=3, activation='relu', name='conv4', padding="SAME")

	# Merge maps #
	network = tflearn.conv_2d(network, nb_filter=1, filter_size=1, activation='linear')

	# Loss and Optimizer #
	network = tflearn.regression(network, optimizer='adam', loss='weak_cross_entropy_2d', name='target')
	
	# Train the model # 
	print("Training the model...")
	model = tflearn.DNN(network, tensorboard_verbose=0,
						checkpoint_path='/home/ger/Desktop/Seg/checkpoints/',
						best_checkpoint_path='/home/ger/Desktop/Seg/checkpoints/')
	model.fit(X, Y, n_epoch=20,
           	  validation_set=0.1,
           	  snapshot_step=100, show_metric=True, run_id='seg_convnet')`

**Comment by aymericdamien** [comment]
How large is your dataset? It may be that all your data are loaded into RAM memory (thus freezing because no RAM left), I am not too sure if your 'dataLoader' loads data on the fly or not, but the line with numpy reshaping will load all your data into RAM memory.
You can check that function for reading images from disk: http://tflearn.org/data_utils/#image-preloader

**Comment by gergf** [comment]
@aymericdamien Thanks for being so fast. I have optimized my loader, as you suggested, and now I can run the net with mini-batches of 20 images 256x256 (not more). I am using niBabel, which allows me to load a slice from a volume without loading the whole image into RAM, I thought this was a good way to feed the net.

Anyways, I feel like the hdf5 is the proper way to feed the nets. I avoided it because I have around 10 GB of images, and for what I read in the documentation, I should "flatten" all the slices of the 3D images and create a .txt which contains in each line one image... but I'll give it a try! 

If you have any advice, please tell me, and feel free to close this issue.
